### PR TITLE
ARROW-11735: [R] Allow Parquet and Arrow Dataset to be optional components

### PR DESCRIPTION
**Pull request by ianmcook**
Not implemented for Windows yet (that's ARROW-11884)

### CODE REVIEW

File: r/R/dataset-partition.R [hunk 1]
```diff
@@ -76,7 +76,9 @@ HivePartitioning$create <- dataset___HivePartitioning
 #' calling `hive_partition()` with no arguments.
 #' @examples
 #' \donttest{
-#' hive_partition(year = int16(), month = int8())
+#' if (arrow_with_dataset()) {
```
Comment (nealrichardson) [inline r/R/dataset-partition.R:79]: Dev roxygen2 has an `@examplesIf` feature that might be better here. If we put the conditional in the example code, it may suggest to the user that they also need to check this every time, and generally they shouldn't.

File: r/data-raw/codegen.R [hunk 2]
```diff
@@ -194,10 +194,16 @@ arrow_exports_cpp <- glue::glue('
 
 {feature_available("arrow")}
 
+{feature_available("dataset")}
```
Comment (nealrichardson) [inline r/data-raw/codegen.R:197]: It might be good to make these bits iterate over `features` rather than copy-paste, just to help with future features.

File: r/tests/testthat/latin1.R [hunk 3]
```diff
@@ -15,7 +15,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
-x <- iconv("Veitingasta�ir", to = "latin1")
+x <- iconv("Veitingasta�ir", to = "latin1")
```
Comment (nealrichardson) [inline r/tests/testthat/latin1.R:18]: Was this change intentional?

File: r/tests/testthat/test-dataset.R [hunk 4]
```diff
@@ -15,6 +15,8 @@
 # specific language governing permissions and limitations
 # under the License.
 
+skip_if_not_available("dataset")
```
Comment (nealrichardson) [inline r/tests/testthat/test-dataset.R:18]: You can skip a whole file like this? Cool.

File: r/tests/testthat/test-python-flight.R [hunk 5]
```diff
@@ -15,6 +15,8 @@
 # specific language governing permissions and limitations
 # under the License.
 
+skip("Temporarily skip this")
```
Comment (nealrichardson) [inline r/tests/testthat/test-python-flight.R:18]: Remove this (also why?)

**Review by nealrichardson — COMMENTED** [review]
In general this looks like the right direction. I'm not certain that the `configure` script changes are exactly what we want but I'll need to think some more about it.

File: r/tests/testthat/test-python-flight.R (hunk 5 above)
Comment (ianmcook) [reply r/tests/testthat/test-python-flight.R:18]: Ah shoot, sorry, I wasn't supposed to commit that 😂

File: r/tests/testthat/latin1.R (hunk 3 above)
Comment (ianmcook) [reply r/tests/testthat/latin1.R:18]: Whoa, no

Comment (nealrichardson) [reply r/tests/testthat/latin1.R:18]: Yeah you need to open this in a text editor that won't reencode to UTF-8 (`vim` may be suited for this), or otherwise set the encoding in the editor to latin1 (or whatever its official name is).

Comment (ianmcook) [reply r/tests/testthat/latin1.R]: Fixed in e894bdd

File: r/configure [hunk 6]
```diff
@@ -26,13 +26,14 @@
...
 # Library settings
-PKG_CONFIG_NAME="arrow parquet arrow-dataset"
+PKG_CONFIG_NAME="arrow"
 PKG_DEB_NAME="(unsuppored)"
 PKG_RPM_NAME="(unsuppored)"
 PKG_BREW_NAME="apache-arrow"
 PKG_TEST_HEADER="<arrow/api.h>"
-# These must be the same order as $(pkg-config --libs arrow-dataset)
-PKG_LIBS="-larrow_dataset -lparquet -larrow"
+PKG_LIBS="-larrow"
```
Comment (nealrichardson) [inline r/configure:34]: I don't understand the specifics but apparently the order of the `-l` libs matters. So we should make sure below that we build up `$PKG_LIBS` so that the order is correct (and note in comments that the order matters).

File: r/configure [hunk 7]
```diff
@@ -69,19 +70,22 @@ fi
 if [ "$INCLUDE_DIR" ] || [ "$LIB_DIR" ]; then
   echo "*** Using INCLUDE_DIR/LIB_DIR"
   PKG_CFLAGS="-I$INCLUDE_DIR $PKG_CFLAGS"
-  PKG_LIBS="-L$LIB_DIR $PKG_LIBS"
+  PKG_DIRS="-L$LIB_DIR $PKG_DIRS"
```
Comment (nealrichardson) [inline r/configure:73]: You've defined `$PKG_DIRS` to be empty above, so there's nothing to append here.

```suggestion
  PKG_DIRS="-L$LIB_DIR"
```

File: r/configure [hunk 8]
```diff
@@ -69,19 +70,22 @@ fi
...
 else
   # Use pkg-config if available and allowed
   pkg-config --version >/dev/null 2>&1
   if [ "$ARROW_USE_PKG_CONFIG" != "false" ] && [ $? -eq 0 ]; then
     PKGCONFIG_CFLAGS=`pkg-config --cflags --silence-errors ${PKG_CONFIG_NAME}`
-    PKGCONFIG_LIBS=`pkg-config --libs --silence-errors ${PKG_CONFIG_NAME}`
+    PKGCONFIG_LIBS=`pkg-config --libs-only-l --silence-errors ${PKG_CONFIG_NAME}`
+    PKGCONFIG_DIRS=`pkg-config --libs-only-L --silence-errors ${PKG_CONFIG_NAME}`
+    # TODO: what about --libs-only-other?
```
Comment (nealrichardson) [inline r/configure:81]: FYI, some of this logic (including stuff further down) should change once https://issues.apache.org/jira/browse/ARROW-6312 happens and the pkgconfig files declare what -l libs they require, but this should be ok for now.

File: r/configure [hunk 9]
```diff
@@ -145,7 +150,7 @@ else
         # TODO: what about non-bundled deps?
         BUNDLED_LIBS=`cd $LIB_DIR && ls *.a`
         BUNDLED_LIBS=`echo $BUNDLED_LIBS | sed -E "s/lib(.*)\.a/-l\1/" | sed -e "s/\\.a lib/ -l/g"`
-        PKG_LIBS="-L$(pwd)/$LIB_DIR $PKG_LIBS $BUNDLED_LIBS"
+        PKG_DIRS="-L$(pwd)/$LIB_DIR $PKG_DIRS"
```
Comment (nealrichardson) [inline r/configure:153]: How would `$PKG_DIRS` be set before here?

Also `$(pwd)` presumably is a bashism we can't use, right? No idea why that hasn't caused us problems before (we even test this on CI).

File: r/configure [hunk 10]
```diff
@@ -182,15 +187,33 @@ if [ $? -eq 0 ] || [ "$UNAME" = "Darwin" ]; then
   # Always build with arrow on macOS
   PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_ARROW"
   # Check for features
-  LIB_DIR=`echo $PKG_LIBS | sed -e 's/ -l.*//' | sed -e 's/^-L//'`
-  grep 'set(ARROW_S3 "ON")' $LIB_DIR/cmake/arrow/ArrowOptions.cmake >/dev/null 2>&1
+  LIB_DIR=`echo $PKG_DIRS | sed -e 's/^-L//'`
+  ARROW_OPTS_CMAKE="$LIB_DIR/cmake/arrow/ArrowOptions.cmake"
+  # Check for Arrow Dataset subcomponent
```
Comment (nealrichardson) [inline r/configure:192]: Per comment above, I think we should prepend -lparquet before -larrow_dataset

File: r/configure [hunk 11]
```diff
@@ -182,15 +187,33 @@ if [ $? -eq 0 ] || [ "$UNAME" = "Darwin" ]; then
...
+  grep 'set(ARROW_DATASET "ON")' $ARROW_OPTS_CMAKE >/dev/null 2>&1
+  if [ $? -eq 0 ]; then
+    PKG_CFLAGS="$PKG_CFLAGS -DARROW_R_WITH_DATASET"
+    PKG_LIBS="-larrow_dataset $PKG_LIBS"
+    # TODO: what if arrow-dataset has a different -L location than arrow?
```
Comment (nealrichardson) [inline r/configure:197]: Not supported. Replace this comment with a comment that just notes we assume that all arrow libs are in the same directory (which in practice they always will be).

File: r/configure [hunk 12]
```diff
@@ -26,13 +26,14 @@
...
-PKG_CONFIG_NAME="arrow parquet arrow-dataset"
+PKG_CONFIG_NAME="arrow"
 PKG_DEB_NAME="(unsuppored)"
 PKG_RPM_NAME="(unsuppored)"
 PKG_BREW_NAME="apache-arrow"
 PKG_TEST_HEADER="<arrow/api.h>"
-# These must be the same order as $(pkg-config --libs arrow-dataset)
-PKG_LIBS="-larrow_dataset -lparquet -larrow"
+PKG_LIBS="-larrow"
+PKG_DIRS=""
```
Comment (nealrichardson) [inline r/configure:35]: Reading through the code, I'm not sure this is needed. I think in every case, you're defining $PKG_DIRS and never want to append an empty string to it.

**Comment by nealrichardson** [comment]
Oh, one other thing: in `inst/build_arrow_static/sh`, you'll want to make ARROW_PARQUET and ARROW_DATASET configurable by env var but default ON, just like ARROW_JEMALLOC. That way, we can make a CI job that sets them to OFF in the env so we can test this.

File: r/R/dataset-partition.R (hunk 1 above)
Comment (ianmcook) [reply r/R/dataset-partition.R:79]: I created ARROW-11849 for us to do that after this feature makes it to CRAN

File: r/data-raw/codegen.R (hunk 2 above)
Comment (ianmcook) [reply r/data-raw/codegen.R]: Done in c8d390d

File: r/configure (hunk 6 above)
Comment (ianmcook) [reply r/configure:34]: Yep, the order in which the code below prepends to `PKG_LIBS` matches the output of `pkg-config --libs-only-l arrow-dataset` (at least on my Mac) which is:
```
-larrow_dataset -lparquet -larrow
```

File: r/configure (hunk 10 above)
Comment (ianmcook) [reply r/configure:192]: I thought so too at first, but when I run
```
pkg-config --libs-only-l arrow-dataset
```
I get `-larrow-dataset` first and `-lparquet` second.

File: r/R/dataset-partition.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/dataset-partition.R:79]: We don't need roxygen2 to be on CRAN in order to use it, the Rd it generates is valid.

Comment (ianmcook) [reply r/R/dataset-partition.R:79]: Right, roxygen2 does not run on CRAN, but I think we should avoid requiring R package contributors to use dev versions of devtools, roxygen2, etc.

File: r/configure (hunk 9 above)
Comment (ianmcook) [reply r/configure:153]: `$(command)` is apparently not a bashism: https://stackoverflow.com/questions/3789894/is-command-substitution-foo-bashism

Comment (ianmcook) [reply r/configure:153]: `PKG_DIRS` contains only the `-L` flags, which are directories. `PKG_LIBS` contain only the `-l` flags, which are package names. Previously these were combined together in `PKG_LIBS`. I separated them because it makes the appending/prepending much more straightforward.

File: r/configure (hunk 8 above)
Comment (ianmcook) [reply r/configure:81]: Cool, thanks for pointing that out.

File: r/configure (hunk 9 above)
Comment (nealrichardson) [reply r/configure:153]: Yeah, I followed that. I just meant that I don't think it's possible for `PKG_DIRS` to be set to anything other than "" before here, so you don't need to append it do you, just set `PKG_DIRS="-L$(pwd)/$LIB_DIR"`?

File: r/configure (hunk 10 above)
Comment (nealrichardson) [reply r/configure:192]: Sorry, my note was ambiguous. The order should be `-larrow_dataset -lparquet`, correct. But if we're prepending them one at a time, we should first prepend parquet, then arrow_dataset. If I'm following correctly, the code here first does if ARROW_DATASET, do `PKG_LIBS="-larrow_dataset $PKG_LIBS"`, then if ARROW_PARQUET, do `PKG_LIBS="-lparquet $PKG_LIBS"`, which would result in `-lparquet -larrow_dataset`

Comment (ianmcook) [reply r/configure:192]: D'oh—I switched from appending to prepending and forgot to change the order 😳 Thank you for catching that.

File: r/configure (hunk 6 above)
Comment (ianmcook) [reply r/configure:34]: Correction: the order was wrong, but this was corrected in b7cdbe26ef6ab9496b2a1f810e84469db9620fe1

File: r/configure (hunk 12 above)
Comment (ianmcook) [reply r/configure]: I confirmed this. Removed in edbfb7e3732cf63e63ba2a7380771a2c56e722f2

File: r/configure (hunk 11 above)
Comment (ianmcook) [reply r/configure]: Done in 3dc2061abc13949249e689c64cbdab810948ec65

File: r/configure (hunk 10 above)
Comment (ianmcook) [reply r/configure]: Fixed in b7cdbe26ef6ab9496b2a1f810e84469db9620fe1

File: r/configure (hunk 9 above)
Comment (ianmcook) [reply r/configure]: Oh, gotcha. I just went through and checked the control flow more carefully, and I agree—`PKG_DIRS` will never be set here. I removed it in 31bd5e9c12023b793a7ef30d6d14f2290698ffc1

File: r/R/dataset-partition.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/dataset-partition.R:79]: I don't feel strongly about not requiring dev roxygen, but it's fine either way.

If you don't want to use examplesIf here, I think you can leave off the `if (arrow_with_dataset())` checks everywhere since it's already inside a `\donttest`

File: r/configure [hunk 13]
```diff
@@ -26,13 +26,13 @@
...
-PKG_CONFIG_NAME="arrow parquet arrow-dataset"
+PKG_CONFIG_NAME="arrow"
 PKG_DEB_NAME="(unsuppored)"
 PKG_RPM_NAME="(unsuppored)"
 PKG_BREW_NAME="apache-arrow"
 PKG_TEST_HEADER="<arrow/api.h>"
-# These must be the same order as $(pkg-config --libs arrow-dataset)
-PKG_LIBS="-larrow_dataset -lparquet -larrow"
+PKG_LIBS="-larrow"
+BUNDLED_LIBS=""
```
Comment (nealrichardson) [inline r/configure:35]: Likewise I don't think you need to set this here, "" is what it will be anyway if it's not set.

File: r/tools/autobrew [hunk 14]
```diff
@@ -48,7 +48,8 @@ fi
 # Hardcode this for my custom autobrew build
 rm -f $BREWDIR/lib/*.dylib
 AWS_LIBS="-laws-cpp-sdk-config -laws-cpp-sdk-transfer -laws-cpp-sdk-identity-management -laws-cpp-sdk-cognito-identity -laws-cpp-sdk-sts -laws-cpp-sdk-s3 -laws-cpp-sdk-core -laws-c-event-stream -laws-checksums -laws-c-common -lpthread -lcurl"
-PKG_LIBS="-L$BREWDIR/lib -lparquet -larrow_dataset -larrow -larrow_bundled_dependencies -lthrift -llz4 -lsnappy -lzstd $AWS_LIBS"
+PKG_LIBS="-lparquet -larrow_dataset -larrow -larrow_bundled_dependencies -lthrift -llz4 -lsnappy -lzstd $AWS_LIBS"
+PKG_DIRS="-L$BREWDIR/lib"
```
Comment (nealrichardson) [inline r/tools/autobrew:52]: We will need to remember to upstream this change when we release to CRAN

File: r/R/dataset-partition.R (hunk 1 above)
Comment (ianmcook) [reply r/R/dataset-partition.R:79]: I initially did that, but apparently as of R 4.0.0:
>`R CMD check --as-cran` now runs `\donttest` examples

(from https://cran.r-project.org/doc/manuals/r-devel/NEWS.html)

Comment (nealrichardson) [reply r/R/dataset-partition.R:79]: It does but I think CRAN doesn't do `--run-donttest` everywhere. Anyway we won't be turning off these features on CRAN except for Solaris so I think it's fine, or worth the risk so as not to pollute the examples with potentially confusing (to users) wrapping.

Comment (ianmcook) [reply r/R/dataset-partition.R:79]: Gotcha. For now, I'll remove the conditional in the example and I'll also change the `\donttest` to `\dontrun`. That way the example checks will pass on Solaris.

File: r/tools/autobrew (hunk 14 above)
Comment (ianmcook) [reply r/tools/autobrew:52]: What is the relationship of this script (`r/tools/autobrew`) to https://github.com/autobrew/scripts/blob/master/apache-arrow?

**Comment by ianmcook** [comment]
> Oh, one other thing: in `inst/build_arrow_static/sh`, you'll want to make ARROW_PARQUET and ARROW_DATASET configurable by env var but default ON, just like ARROW_JEMALLOC. That way, we can make a CI job that sets them to OFF in the env so we can test this.

Done in 8c5b235dddac47833b229323b4a216049eae5fea

File: r/configure (hunk 13 above)
Comment (ianmcook) [reply r/configure]: Done in 13edc1478e5003e3a2b3dc27612c6d85f2aa9269

File: r/R/dataset-partition.R (hunk 1 above)
Comment (ianmcook) [reply r/R/dataset-partition.R]: Done in 6918441bba2b0ca09f2103eaf5631adc0a076a6c

File: r/tools/autobrew (hunk 14 above)
Comment (nealrichardson) [reply r/tools/autobrew:52]: It's based on an earlier version of Jeroen's autobrew, and it's closest now to how he actually builds the C++ dependencies. It looks like the new autobrew script is based on some artifacts that are bundled up in a separate process (see https://github.com/autobrew/bundler/blob/master/apache-arrow.sh).

File: r/R/dataset-partition.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/dataset-partition.R]: Thanks, please also remove the conditionals from the other examples. I don't feel strongly about `\donttest` vs `\dontrun` (CRAN complains about dontrun examples on initial submission, which is why they were switched to donttest in the first place, but that's not our concern now) so that's fine, but FWIW `\donttest` is safe on Solaris: [we currently "pass" Solaris checks](https://www.r-project.org/nosvn/R.check/r-patched-solaris-x86/arrow-00check.html) with the fake arrow-without-arrow build.

Comment (ianmcook) [reply r/R/dataset-partition.R]: Done in 26753e4e0ed08739467588cfb3ddcfe09f5aebcc. I will change the `\dontrun`s back to `\donttest` (or remove them altogether if possible) when I implement `@examplesIf`, as noted in ARROW-11849

File: r/tools/autobrew (hunk 14 above)
Comment (ianmcook) [reply r/tools/autobrew:52]: I created ARROW-11861 to remind us of this.

**Comment by ianmcook** [comment]
@github-actions crossbow submit -g r

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build

**Comment by ianmcook** [comment]
@nealrichardson do you want `LIBARROW_MINIMAL` to toggle Dataset and Parquet? I assume no.

File: dev/tasks/tasks.yml [hunk 15]
```diff
@@ -1789,6 +1789,18 @@ tasks:
       r_image: r-base
       r_tag: 3.6-opensuse42
       not_cran: "TRUE"
+  
+  test-r-minimal-build:
```
Comment (ianmcook) [inline dev/tasks/tasks.yml:1793]: TODO: Use a different word because "minimal" already has a specific meaning (`LIBARROW_MINIMAL`) that's not this

**Comment by nealrichardson** [comment]
> @nealrichardson do you want `LIBARROW_MINIMAL` to toggle Dataset and Parquet? I assume no.

Correct, we don't want to change that meaning (yet).

File: dev/tasks/tasks.yml (hunk 15 above)
Comment (nealrichardson) [reply dev/tasks/tasks.yml:1793]: But it is minimal in the sense that the cpp minimal build in here is. I think it's fine.

File: dev/tasks/r/azure.linux.yml [hunk 16]
```diff
@@ -47,6 +47,9 @@ jobs:
         export R_ORG={{ r_org }}
         export R_IMAGE={{ r_image }}
         export R_TAG={{ r_tag }}
+        export ARROW_DATASET={{ arrow_dataset|default("ON") }}
+        export ARROW_PARQUET={{ arrow_parquet|default("ON") }}
```
Comment (nealrichardson) [inline dev/tasks/r/azure.linux.yml:51]: Unfortunately this didn't pass through all the way: https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=1721&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=240

I believe what you instead want is to pass them into docker-compose below, something like `docker-compose -e ARROW_DATASET={{ arrow_dataset|default("ON") }} -e ARROW_PARQUET={{ arrow_parquet|default("ON") }} -e ARROW_S3={{ arrow_s3|default("ON") }} r`

Comment (ianmcook) [reply dev/tasks/r/azure.linux.yml:51]: Got it, thanks. Let's see if it works to add them to `.env` (066ce065124cf7dcaa08225cae83177594da3965) and if not then I'll do it with `-e`.

Comment (ianmcook) [reply dev/tasks/r/azure.linux.yml:51]: Oops, need to append not overwrite. Fixed in 318371fbaa07f1b6fcdf71f58d7d5460a931da13

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build

File: dev/tasks/r/azure.linux.yml (hunk 16 above)
Comment (nealrichardson) [reply dev/tasks/r/azure.linux.yml:51]: I think the only danger to appending to the .env file is that if you ran this locally, wouldn't it modify the file locally?

Comment (ianmcook) [reply dev/tasks/r/azure.linux.yml:51]: Ok, if folks might be running these commands locally then I can see how that would cause damage.
Looks like it didn't work anyhow. I will investigate.

Comment (nealrichardson) [reply dev/tasks/r/azure.linux.yml:51]: Ah yeah, I think it's the same problem--note (for comparison) that `R_ORG` et al. are in the .env file, and they're environment variables for docker-compose, not passed into the docker container.

The fact that `ARROW_R_DEV` is in the env file, and then passed through in docker-compose.yml to the build environments, may just be due to my ignorance of all of this at the time I added it.

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build

File: dev/tasks/r/azure.linux.yml (hunk 16 above)
Comment (ianmcook) [reply dev/tasks/r/azure.linux.yml]: Ok, thanks. I tried a few different things just for kicks, then used `-e` per your suggestion in 9479c6ce079c7230b0e290dcb5658436d1d62fde

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build

**Comment by ianmcook** [comment]
@github-actions crossbow submit -g r

File: dev/tasks/r/azure.linux.yml (hunk 16 above)
Comment (ianmcook) [reply dev/tasks/r/azure.linux.yml]: The build looks good after 9479c6c:
https://dev.azure.com/ursacomputing/crossbow/_build/latest?definitionId=1&branchName=actions-184-azure-test-r-minimal-build

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build

**Comment by ianmcook** [comment]
@github-actions crossbow submit -g r

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-without-dataset-parquet-s3

**Comment by ianmcook** [comment]
The environment variables are being set in the container now, and Arrow is building with Dataset and Parquet off, but S3 is remaining on despite the environment variable.

**Comment by nealrichardson** [comment]
S3 is on because linuxlibs.R is turning it on: https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=1741&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=24

I think you'll find the issue pretty quickly but LMK if you need a pointer.

File: dev/tasks/tasks.yml [hunk 17]
```diff
@@ -1789,6 +1789,18 @@ tasks:
       r_image: r-base
       r_tag: 3.6-opensuse42
       not_cran: "TRUE"
+  
+  test-r-without-dataset-parquet-s3:
```
Comment (nealrichardson) [inline dev/tasks/tasks.yml:1793]: I actually liked the "minimal" name, matches the cpp minimal build example we test. If you're worried about confusion with the `LIBARROW_MINIMAL` env var (I'm not personally), we could try to make that var be truly minimal. I think currently we only ever check for whether it is set to `false`, to mean "build all the compression libs etc.", but we don't check if it is `true`. So we could make `true` be truly minimal, without these optional features, and the default (unset) means the current default of all features on but not compression and S3.

On an unrelated note, you should add this crossbow job to the `r` and `nightly` groups so that it gets run regularly.

Comment (ianmcook) [reply dev/tasks/tasks.yml:1793]: > S3 is on because linuxlibs.R is turning it on: https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=1741&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=6c939d89-0d1a-51f2-8b30-091a7a82e98c&l=24
> 
> I think you'll find the issue pretty quickly but LMK if you need a pointer.

Looks like `configure` is setting `LIBARROW_MINIMAL=false` because `NOT_CRAN` is `true`:
```sh
if [ "${LIBARROW_MINIMAL}" = "" ] && [ "${NOT_CRAN}" = "true" ]; then
  LIBARROW_MINIMAL=false; export LIBARROW_MINIMAL
fi
```
and then `linuxlibs.r` sets `arrow_s3` to `TRUE`:
```r
arrow_s3 <- toupper(Sys.getenv("ARROW_S3")) == "ON" || tolower(Sys.getenv("LIBARROW_MINIMAL")) == "false"
```
I will try to solve this by explicitly setting `LIBARROW_MINIMAL=true` in this build.

Comment (ianmcook) [reply dev/tasks/tasks.yml]: > I actually liked the "minimal" name, matches the cpp minimal build example we test. If you're worried about confusion with the `LIBARROW_MINIMAL` env var (I'm not personally), we could try to make that var be truly minimal. I think currently we only ever check for whether it is set to `false`, to mean "build all the compression libs etc.", but we don't check if it is `true`. So we could make `true` be truly minimal, without these optional features, and the default (unset) means the current default of all features on but not compression and S3.
> 
> On an unrelated note, you should add this crossbow job to the `r` and `nightly` groups so that it gets run regularly.

For now I think I'll just set `LIBARROW_MINIMAL=true` so the "minimal" name is not misleading. Later we might consider whether an explicit `LIBARROW_MINIMAL=true` should toggle Dataset and Parquet off.

File: dev/tasks/conda-recipes/r-arrow/configure.win [hunk 18]
```diff
@@ -2,7 +2,7 @@
 
 set -euxo pipefail
 
-echo "PKG_CPPFLAGS=-DNDEBUG -I\"${LIBRARY_PREFIX}/include\" -I\"${PREFIX}/include\" -DARROW_R_WITH_ARROW -DARROW_R_WITH_S3" > src/Makevars.win
+echo "PKG_CPPFLAGS=-DNDEBUG -I\"${LIBRARY_PREFIX}/include\" -I\"${PREFIX}/include\" -DARROW_R_WITH_ARROW -DARROW_R_WITH_PARQUET -DARROW_R_WITH_DATASET -DARROW_R_WITH_S3" > src/Makevars.win
```
Comment (ianmcook) [inline dev/tasks/conda-recipes/r-arrow/configure.win:5]: @nealrichardson is this another change that we will need to upstream at the next release?

**Comment by ianmcook** [comment]
@github-actions crossbow submit -g r

File: dev/tasks/conda-recipes/r-arrow/configure.win (hunk 18 above)
Comment (nealrichardson) [reply dev/tasks/conda-recipes/r-arrow/configure.win:5]: Yes. You could make the change now in fact, nothing would be hurt by having extra -D flags on. https://github.com/conda-forge/r-arrow-feedstock (FYI @xhochy)

**Comment by nealrichardson** [comment]
Good catch about the Windows builds. Could you make followup JIRAs for:

1. configure.win should determine Arrow C++ library capabilities dynamically (like we do in configure)
2. `LIBARROW_MINIMAL=true` should entail that all optional features are `OFF`

**Comment by ianmcook** [comment]
> 1. configure.win should determine Arrow C++ library capabilities dynamically (like we do in configure)

ARROW-11884

> 2. `LIBARROW_MINIMAL=true` should entail that all optional features are `OFF`

ARROW-11885

File: dev/tasks/conda-recipes/r-arrow/configure.win (hunk 18 above)
Comment (ianmcook) [reply dev/tasks/conda-recipes/r-arrow/configure.win:5]: Done https://github.com/conda-forge/r-arrow-feedstock/pull/30

**Comment by nealrichardson** [comment]
Sorry, I just merged the dataset mutate() PR which added a cpp function. Can you please rebase and update the annotation on that function? Otherwise this is looking good, will merge.

**Comment by ianmcook** [comment]
> Sorry, I just merged the dataset mutate() PR which added a cpp function. Can you please rebase and update the annotation on that function? Otherwise this is looking good, will merge.

Done

File: dev/tasks/tasks.yml (hunk 17 above)
Comment (ianmcook) [reply dev/tasks/tasks.yml]: The `r` and `nightly` groups are defined with wildcards that will match the name I used for this test.

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-r-minimal-build

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-minimal-build

File: dev/tasks/r/azure.linux.yml [hunk 19]
```diff
@@ -50,7 +50,12 @@ jobs:
         # we have to export this (right?) because we need it in the build env
         export ARROW_R_DEV={{ not_cran }}
         # Note that ci/scripts/r_test.sh sets NOT_CRAN=true if ARROW_R_DEV=TRUE
-        docker-compose run r
+        docker-compose run \
+          -e ARROW_DATASET={{ arrow_dataset|default("ON") }} \
+          -e ARROW_PARQUET={{ arrow_parquet|default("ON") }} \
+          -e ARROW_S3={{ arrow_s3|default("ON") }} \
```
Comment (nealrichardson) [inline dev/tasks/r/azure.linux.yml:56]: This should work too right? That way by default we're testing the default, unset versions.

```suggestion
          -e ARROW_DATASET={{ arrow_dataset|default("") }} \
          -e ARROW_PARQUET={{ arrow_parquet|default("") }} \
          -e ARROW_S3={{ arrow_s3|default("") }} \
```

I'll try it and revert if it's no good.

**Review by nealrichardson — APPROVED** [review]
This is great, thanks for doing this.

I'm going to try one more thing with the crossbow jobs, will revert if it's no good, and then merge.

File: dev/tasks/r/azure.linux.yml (hunk 19 above)
Comment (ianmcook) [reply dev/tasks/r/azure.linux.yml]: I think so

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rstudio* test-r-minimal-build

File: dev/tasks/conda-recipes/r-arrow/configure.win (hunk 18 above)
Comment (ianmcook) [reply dev/tasks/conda-recipes/r-arrow/configure.win:5]: Merged in https://github.com/conda-forge/r-arrow-feedstock/commit/ed01e9badf1cd831d5201d6f9b35dcf01c873cf6

File: r/tools/autobrew (hunk 14 above)
Comment (ianmcook) [reply r/tools/autobrew:52]: Turns out we also needed to add  `-DARROW_R_WITH_PARQUET` and `-DARROW_R_WITH_DATASET` to the cflags in ttps://github.com/autobrew/scripts/blob/master/apache-arrow (ARROW-12604)
